Examples/ObjectiveC/MultipleShapesExample.m

```diff
@@ -7,61 +7,80 @@ @interface MultipleShapesExample () <MGLMapViewDelegate>
 @property (nonatomic) MGLMapView *mapView;
 @end
 
+@interface MGLStyle (LayerExtensions)
+- (void)addLinesFrom:(MGLShapeSource *)source;
+- (void)addPointsFrom:(MGLShapeSource *)source;
+- (void)addPolygonsFrom:(MGLShapeSource *)source;
+@end
+
 @implementation MultipleShapesExample
 
 - (void)viewDidLoad {
     [super viewDidLoad];
-    
+
+    // Configure a map view centered on Washington, D.C.
     self.mapView = [[MGLMapView alloc] initWithFrame:self.view.bounds styleURL:[MGLStyle lightStyleURL]];
     self.mapView.autoresizingMask = UIViewAutoresizingFlexibleWidth | UIViewAutoresizingFlexibleHeight;
-    self.mapView.tintColor = [UIColor darkGrayColor];
-    
     [self.mapView setCenterCoordinate:CLLocationCoordinate2DMake(38.897435, -77.039679) zoomLevel:12 animated:NO];
     self.mapView.delegate = self;
     
     [self.view addSubview:self.mapView];
 }
 
 - (void)mapView:(MGLMapView *)mapView didFinishLoadingStyle:(MGLStyle *)style {
-    
-    // Parse the GeoJSON data.
-    dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_DEFAULT, 0), ^{
-        NSString *path = [[NSBundle mainBundle] pathForResource:@"metro-line" ofType:@"geojson"];
-        
-        NSData *data = [NSData dataWithContentsOfFile:path];
-        
-        dispatch_async(dispatch_get_main_queue(), ^{
-            [self drawShapeCollection:data];
-        });
-    });
+    // Since we know this file exists within this project, we'll force unwrap its value.
+    // If this data was coming from an external server, we would want to perform
+    // proper error handling for a web request/response.
+    NSURL *url = [NSURL fileURLWithPath:[NSBundle.mainBundle pathForResource:@"metro-line" ofType:@"geojson"]];
+
+    // Create a shape source and register it with the map style.
+    MGLShapeSource *source = [[MGLShapeSource alloc] initWithIdentifier:@"metro-line" URL:url options:nil];
+    [style addSource:source];
+
+    // Add different line, point, and polygon shapes to the map style.
+    [style addLinesFrom:source];
+    [style addPointsFrom:source];
+    [style addPolygonsFrom:source];
 }
 
-- (void)drawShapeCollection:(NSData *)data {
-    
-    // Use [MGLShape shapeWithData:encoding:error:] to create a MGLShapeCollectionFeature from GeoJSON data.
-    MGLShape *feature = [MGLShape shapeWithData:data encoding:NSUTF8StringEncoding error:NULL];
-    
-    // Create source and add it to the map style.
-    MGLShapeSource *source = [[MGLShapeSource alloc] initWithIdentifier:@"transit" shape:feature options:nil];
-    [self.mapView.style addSource:source];
-    
-    // Create station style layer.
+@end
+
+@implementation MGLStyle (LayerExtensions)
+
+- (void)addLinesFrom:(MGLShapeSource *)source {
+    // Configure a line style layer to represent a rail line, filtering out all data from the
+    // source that is not of `Rail line` type.
```

Comment:
Should we say that `TYPE` is an attribute key, just so people know where `Rail line` and the other types are coming from?